### ISSUE TITLE
Softmax is everywhere!

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#179]
# Softmax 是什么

softmax 是一个把“一组实数”转换成“概率分布”的函数，在深度学习（尤其分类模型、LLM）中极其核心。

我分三步说清楚：  
1）softmax 是什么（公式 + 性质）  
2）在深度学习/LLM里用它干嘛  
3）为什么它这么重要而且几乎到处都有

---

## 1. softmax 是什么？

给你一组实数向量（通常叫 logits）：

\[
z = (z_1, z_2, \dots, z_K)
\]

softmax 把它变成一组“看上去像概率”的数：

\[
\text{softmax}(z)_i = \frac{e^{z_i}}{\sum_{j=1}^{K} e^{z_j}},\quad i = 1,\dots,K
\]

有几个关键性质：

1. **每个输出都是正数**：  
   因为 \(e^{z_i} > 0\)。

2. **总和为 1**：  
   \(\sum_i \text{softmax}(z)_i = 1\)，所以可以当作“多类别上的概率分布”。

3. **单调、相对差异有放大效果**：  
   - 如果 \(z_i\) 比 \(z_j\) 大很多，那 \(\text{softmax}(z)_i\) 会显著大于 \(\text{softmax}(z)_j\)；  
   - 放大“最大值”的优势，使得分布更加“尖锐”。

4. **对整体平移不敏感（平移不变性）**：  
   - 对所有 \(z_i\) 加同一个常数 \(c\)，输出不变：  
     \(\text{softmax}(z + c) = \text{softmax}(z)\)  
   - 这在数值稳定性上很重要（经常会用 `z - max(z)` 来避免溢出）。

---

### 2.1 多分类任务的输出层

对于一个 K 类分类问题（比如识别图片是猫/狗/鸟/...），网络最后一层通常输出一个长度为 K 的向量 logits：

\[
z = W x + b
\]

再通过 softmax：

\[
p_i = \text{softmax}(z)_i
\]

得到每一类的概率 \(p_i\)。然后用 **交叉熵损失（cross entropy loss）** 与真实标签计算损失来训练：

\[
\text{Loss} = - \sum_{i=1}^K y_i \log p_i
\]

其中 \(y_i\) 是 one-hot 标签。

所以在“分类任务里”，softmax 的位置就是：

> 最后一层，把模型的原始打分（logits）变成类别概率，用来：  
> 1）给出预测结果  
> 2）配合交叉熵做梯度学习

---

### 2.2 LLM（大语言模型）里的 softmax

LLM 其实本质上也是一个“超大规模分类模型”，只不过它的类别是“词表里的所有 token”。

- 输入是上下文 token 序列；  
- 模型最后对“下一个 token”输出一个 logits 向量（维度 = vocab size，比如 50k、100k 等）；  
- 对这个 logits 做 softmax 得到对每个 token 的概率：

\[
p(\text{token } i | \text{context}) = \text{softmax}(z)_i
\]

LLM 用 softmax 主要在两个地方：

1. **训练阶段**  
   - 目标是让“真实下一 token”的概率最大，仍旧用交叉熵损失：  
     \(-\log p(\text{真实 token})\)。  
   - softmax + cross entropy 让梯度传播稳定、有效。

2. **生成阶段（推理）**  
   - 每一步预测时，softmax 得到下一个 token 的概率分布；  
   - 再按一定采样策略（greedy / top-k / top-p / temperature 等）从这个分布中选取下一个 token。  
   - 例如 temperature 调节，就是对 logits / softmax 分布进行“变尖/变平”的控制。

---

### 2.3 自注意力机制（self-attention）里的 softmax

除了输出层，Transformer/LLM 的**内部结构**也大量用到 softmax。

在自注意力（self-attention）里：

1. 先计算 query 和 key 的点积得到注意力分数矩阵 \(S\)：
   \[
   S = \frac{QK^T}{\sqrt{d_k}}
   \]
2. 对每一行（每个 query 对所有 key 的分数）做 softmax：
   \[
   A = \text{softmax}(S)
   \]
   得到注意力权重矩阵 \(A\)。

3. 用这些权重对 value 做加权求和：
   \[
   \text{Attention}(Q, K, V) = A V
   \]

这里 softmax 的作用是：

- 把任意实数分数 \(S_{ij}\) 转成非负、和为 1 的权重；
- 表示“在所有单词中，当前这个单词应该关注谁、关注多少”。

没有 softmax，自注意力就没法自然形成概率型的“注意分布”。

---

## 3. 为什么软最大（softmax）如此重要？

总结一下它被广泛使用的原因：

### 3.1 把“打分”变成“概率”

- 模型内部喜欢输出“线性打分”（logits），因为这样便于线性/仿射变换和梯度计算；
- 但我们在人类层面、在损失函数层面，通常需要“概率”的概念（分类、决策、采样、对数似然等）。  
- softmax 完成了从“任意实数向量”到“概率分布”的桥接。

### 3.2 与交叉熵组合，梯度性质很好

softmax + cross entropy 这一对，在数学上：

- 有简洁的梯度形式；
- 在数值上相对稳定；
- 在优化上效果好，被实证为深度学习里“多分类问题”的事实标准。

这种“好用 + 标准”的组合，导致几乎所有分类类模型（包括语言模型的 token 预测）都采用它。

### 3.3 允许表达“相对差异”，不仅是绝对值

因为 softmax 基于指数和归一化：

- 只关心各分量之间的相对大小；  
- 调整 logits 的差异，就能控制分布的“尖锐程度”。  
  - 差异大 → 分布更尖锐（更确定，类似 one-hot）  
  - 差异小 → 分布更平（表示不确定）

这对于模型表达“信心”非常重要：  
- LLM 的温度（temperature）调节，就是在 logits 上做缩放，然后再 softmax，从而控制“随机性 / 创造力”。

### 3.4 为注意力机制提供“概率型权重”

- 自注意力、自注意力+多头注意力，是 Transformer/LLM 成功的关键。  
- softmax 将任意分数转为“注意力权重概率分布”，自然且可微，利于训练。  
- 没有 softmax，就难以用一个统一、平滑、可微的方式来表达“对不同元素的权重分配”。

---

## 4. 小结一句话版

- **softmax 是一个把一组任意实数转换成概率分布的函数。**  
- 在深度学习和 LLM 中，它主要负责：
  - 把模型输出的 logits 转成类别/词的概率（输出层）；  
  - 在注意力机制中，把相似度分数转成注意力权重。  
- 它之所以“无处不在”，是因为：  
  - 它完美地连接了“可微分的模型打分”与“概率/决策/采样”；  
  - 与交叉熵损失的组合在数学性质、数值稳定性和训练效果上都非常优秀；  
  - 是注意力机制和大语言模型的基础组件之一。

# 简易实现 softmax

下面给你两个极简示例：一个是普通分类，一个是“下一个 token 预测”（LLM 的核心步骤）。都各用一版 PyTorch 和一版 NumPy。

---

## 一、分类任务中的 softmax

设想：有一个 3 类分类任务（如 猫/狗/鸟），模型最后输出 logits，然后用 softmax 得到每一类的概率。

### 1. PyTorch 版本

```python
import torch
import torch.nn.functional as F

# 假设 batch_size = 2，每个样本有 3 类的 logits
logits = torch.tensor([[1.2, 0.3, -0.5],
                       [0.1, 2.0, 0.0]])   # shape: (2, 3)

# 用 softmax 把 logits 变成概率
probs = F.softmax(logits, dim=-1)
print("probabilities:\n", probs)
print("row sums:", probs.sum(dim=-1))  # 每行和应该是 1

# 假设真实标签（类别索引）
targets = torch.tensor([0, 1])         # 第一个样本真是类0，第二个样本真是类1

# 典型的训练损失：cross entropy（内部自动做 log_softmax）
loss = F.cross_entropy(logits, targets)
print("cross entropy loss:", loss.item())
```

### 2. NumPy 版本（手算 softmax）

```python
import numpy as np

logits = np.array([[1.2, 0.3, -0.5],
                   [0.1, 2.0, 0.0]])  # shape: (2, 3)

# 数值稳定版 softmax：减去每行最大值再 exp
def softmax(x, axis=-1):
    x_max = np.max(x, axis=axis, keepdims=True)
    e_x = np.exp(x - x_max)
    return e_x / np.sum(e_x, axis=axis, keepdims=True)

probs = softmax(logits, axis=-1)
print("probabilities:\n", probs)
print("row sums:", probs.sum(axis=-1))  # 每行和是 1
```

---

## 二、LLM “下一个 token 预测”中的 softmax

简化场景：模型最后输出一个长度为 vocab_size 的向量 logits（对每个 token 一个分数），用 softmax 得到概率分布，然后从中采样一个 token 作为下一个输出。

### 1. PyTorch 版本

```python
import torch
import torch.nn.functional as F

vocab_size = 5  # 假设词表只有 5 个 token（id: 0~4）

# 模型对“下一个 token”输出的 logits（这里随机模拟）
logits = torch.tensor([1.0, 0.5, -0.5, 2.0, 0.0])  # shape: (5,)

# 用 softmax 得到对每个 token 的概率
probs = F.softmax(logits, dim=-1)
print("token probabilities:", probs)

# 按概率分布采样一个 token 作为模型输出
next_token = torch.multinomial(probs, num_samples=1)
print("sampled next token id:", next_token.item())
```

如果你想控制“创造性”（temperature），可以在 softmax 前先缩放 logits：

```python
temperature = 0.7  # <1 更确定，>1 更随机
probs = F.softmax(logits / temperature, dim=-1)
```

### 2. NumPy 版本（模拟 LLM 下一个 token）

```python
import numpy as np

vocab_size = 5
logits = np.array([1.0, 0.5, -0.5, 2.0, 0.0])

def softmax(x):
    x_max = np.max(x)
    e_x = np.exp(x - x_max)
    return e_x / e_x.sum()

probs = softmax(logits)
print("token probabilities:", probs)

# 按概率分布随机采样一个 token id
next_token = np.random.choice(len(probs), p=probs)
print("sampled next token id:", next_token)
```

---
